### PR TITLE
Update output and template generation to type removal in ES API

### PR DESCRIPTION
**Pull request by urso**
- update template generation to not include the type name if ES>=7
- do not set _type in bulk request if ES>=7

Companion PR for 6.7 is #10140.

Both PRs should be merged after both are green and I had a chance to verify upgrading/installing between a few Beats/ES combinations works seamlessly.

### CODE REVIEW

File: libbeat/template/template.go [hunk 1]
```diff
@@ -182,116 +181,156 @@ func (t *Template) GetPattern() string {
...
-		indexSettings.Put("number_of_routing_shards", defaultNumberOfRoutingShards)
+	if len(source) > 0 {
+		mapping["_source"] = source
 	}
 
-	mappingName := "_doc"
-	major := t.esVersion.Major
+	major := esVersion.Major
 	switch {
+	case major == 2:
```
[the remark this answers is not in the document]
Comment (urso) [reply libbeat/template/template.go]: Code for 2.x is minimal. I'd prefer to keep it for now and maybe cleanup in a followup PR.

File: libbeat/template/load.go [hunk 2]
```diff
@@ -32,7 +32,6 @@ import (
 // TemplateLoader is a subset of the Elasticsearch client API capable of
 // loading the template.
 type ESClient interface {
-	LoadJSON(path string, json map[string]interface{}) ([]byte, error)
```
Comment (ph) [inline libbeat/template/load.go]: ++ to remove that.

File: libbeat/template/template.go (hunk 1 above)
[the remark this answers is not in the document]
Comment (ph) [reply libbeat/template/template.go]: It is EOL and I sure we are probably broken with it. I think we should remove it, but we can do it in another PR

File: libbeat/template/template.go [hunk 3]
```diff
@@ -184,116 +183,156 @@ func (t *Template) GetPattern() string {
...
+	if len(a) == 0 {
+		return b, nil
+	}
 
-		// Append is only allowed to add fields, not overwrite
-		for _, key := range appendFieldKeys {
-			if fields.HasNode(key) {
-				return nil, fmt.Errorf("append_fields contains an already existing key: %s", key)
-			}
+	// check for duplicats
```
Comment (ph) [inline libbeat/template/template.go]: s/duplicats/duplicates/

File: libbeat/template/template.go [hunk 4]
```diff
@@ -184,116 +183,156 @@ func (t *Template) GetPattern() string {
...
+	for _, k := range b.GetKeys() {
+		if a.HasNode(k) {
+			return nil, fmt.Errorf("append_fields contains an already existing key: %s", k)
```
Comment (ph) [inline libbeat/template/template.go]: s/append_fields/concat_fields?

File: libbeat/template/template.go [hunk 5]
```diff
@@ -184,116 +183,156 @@ func (t *Template) GetPattern() string {
...
 
-	return basicStructure
+	indexSettings.DeepUpdate(userSettings)
+	return indexSettings
 }
 
-func appendFields(fields, appendFields common.Fields) (common.Fields, error) {
-	if len(appendFields) > 0 {
-		appendFieldKeys := appendFields.GetKeys()
+func concatFields(a, b common.Fields) (common.Fields, error) {
```
Comment (kvch) [inline libbeat/template/template.go]: I think this function is general enough to be moved to `libbeat/common/fields.go`.

Comment (urso) [reply libbeat/template/template.go]: moved + added cleaned up fields_test.go a little.

**Comment by ruflin** [comment]
jenkins, test this

**Comment by urso** [comment]
- update template generation to not include the type name if ES>=7
- do not set _type in bulk request if ES>=7

**Comment by SubodhChettri** [comment]
Let's break OpenSearch!!

**Pull request merged**